File: solutions/javascript/sublist/1/sublist.js
export class List extends Array {
  constructor(elements = []) {
    if (elements.length > 0) super(...elements);
    else super();
  }

  compare(list) {
    if (this.length === 0 && list.length == 0) return "EQUAL";
    if (this.length === 0 && list.length > 0) return "SUBLIST";
    if (this.length > 0 && list.length === 0) return "SUPERLIST";

    if (this.length === list.length) {
      return this.every((elem, idx) => elem === list[idx])
        ? "EQUAL"
        : "UNEQUAL";
    } else if (this.length < list.length) {
      for (let i = 0; i <= list.length - this.length; i++) {
        if (
          list
            .slice(i, i + this.length)
            .every((elem, idx) => elem === this[idx])
        )
          return "SUBLIST";
      }
    } else if (this.length > list.length) {
      for (let i = 0; i <= this.length - list.length; i++) {
        if (
          this.slice(i, i + list.length).every(
            (elem, idx) => elem === list[idx]
          )
        )
          return "SUPERLIST";
      }
    }

    return "UNEQUAL";
  }
}

File: solutions/c/sublist/3/sublist.c
#include "sublist.h"

#include <stdbool.h>
#include <string.h>

comparison_result_t check_lists(int *list_to_compare, int *base_list,
                                size_t list_to_compare_element_count,
                                size_t base_list_element_count) {
	// Handle Degenerate Cases of Empty Lists
	if (list_to_compare_element_count == 0 && base_list_element_count > 0)
		return SUBLIST;

	if (base_list_element_count == 0 && list_to_compare_element_count > 0)
		return SUPERLIST;

	bool base_is_larger = true;

	// If they are equal, we can just compare with memcmp
	if (list_to_compare_element_count == base_list_element_count) {
		if (memcmp(list_to_compare, base_list,
		           sizeof(int) * list_to_compare_element_count) == 0)
			return EQUAL;
		else
			return UNEQUAL;
	} else {
		base_is_larger =
		    list_to_compare_element_count < base_list_element_count;
	}

	size_t larger_count = base_is_larger ? base_list_element_count
	                                     : list_to_compare_element_count;
	int *larger = base_is_larger ? base_list : list_to_compare;
	size_t smaller_count = base_is_larger ? list_to_compare_element_count
	                                      : base_list_element_count;
	int *smaller = base_is_larger ? list_to_compare : base_list;

	for (size_t s = 0, l = 0; l < larger_count; l++) {
		if (larger[l] == smaller[s]) {
			if (s == smaller_count - 1)
				return base_is_larger ? SUBLIST : SUPERLIST;
			else
				s++;
		} else if (s > 0) {
			// Backtrack is case of failed partial match
			l -= s;
			s = 0;
		}
	}

	return UNEQUAL;
}